### DISCUSSION TITLE
MacOSX 14.3 / M3 Pro Chip: Import Error ODBC Driver "no such file" but everything is installed!

### DISCUSSION THREAD

**Discussion by sm-ffm** [mkleehammer/pyodbc#1342] [Q&A] [answered]
Hello,
I have the following Import Error:

>  import pyodbc
ImportError: dlopen(/Users/XXXX/.conda/envs/XXXX/lib/python3.9/site-packages/pyodbc.cpython-39-darwin.so, 0x0002): Library not loaded: /usr/local/opt/unixodbc/lib/libodbc.2.dylib
  Referenced from: <9820BF4A-872E-3566-A1CA-2902C9CCBCE4> /Users/XXXX/.conda/envs/XXXX/lib/python3.9/site-packages/pyodbc.cpython-39-darwin.so
  Reason: tried: '/usr/local/opt/unixodbc/lib/libodbc.2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/opt/unixodbc/lib/libodbc.2.dylib' (no such file), '/usr/local/opt/unixodbc/lib/libodbc.2.dylib' (no such file), '/usr/local/lib/libodbc.2.dylib' (no such file), '/usr/lib/libodbc.2.dylib' (no such file, not in dyld cache)

I have tried a lot of things including:

- installing unixODBC Drivers first before installing pyodbc
- installing Microsoft SQL Drivers as written here: https://learn.microsoft.com/en-gb/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver16
The Installation as on Microsofts Page written failed in first place due to terminal could not understand "brew" - I figured out a solution using an edited .zshrc File. And then the brew commands worked.
All is installed now, but the above Error still is occurring.

In Terminal with Command `odbcinst -j` I get following information:

> unixODBC 2.3.12
DRIVERS............: /opt/homebrew/etc/odbcinst.ini
SYSTEM DATA SOURCES: /opt/homebrew/etc/odbc.ini
FILE DATA SOURCES..: /opt/homebrew/etc/ODBCDataSources
USER DATA SOURCES..: /Users/XXXX/.odbc.ini
SQLULEN Size.......: 8
SQLLEN Size........: 8
SQLSETPOSIROW Size.: 8

In Terminal with Command `cat /opt/homebrew/etc/odbcinst.ini` I get following information:

> [ODBC Driver 18 for SQL Server]
Description=Microsoft ODBC Driver 18 for SQL Server
Driver=/opt/homebrew/lib/libmsodbcsql.18.dylib
UsageCount=1

So, I think everything needed should be installed. But I do not know how to fix that "Directory" Issue.
How can I define the correct Folder to Importing Procedure for pyodbc?
Thank you.

**Comment by v-chojas** [comment]
Could you try `isql -v -k '<your connection string here>'` and see if that connects successfully?

Also, are you using arm64 or amd64 Python?

**Reply by sm-ffm** [reply]
Hi, thanks for your reply. I have done that with following result:

> isql -v -k 'DRIVER={ODBC Driver 18 for SQL Server};SERVER=tcp:XX.XXX.XXX.XXX,XXXXX;DATABASE=XXX;UID=XXXXXXXXXXXXX;TrustServerCertificate=yes;PWD=XXXXXXXX'
+---------------------------------------+
| Connected!                                                |
|                                                                     |
| sql-statement                                            |
| help [tablename]                                       |
| echo [string]                                              |
| quit                                                             |
|                                                                    |
+---------------------------------------+
SQL> 

so it worked actually.... I think the ODBC driver is working fine. But not with pyODBC ... Do you have any ideas how to fix that?

**Reply by sm-ffm** [reply]
Concerning your second question:

> python3 --version
Python 3.8.8
(base) MacBook-Pro-2:~ XXXXXX$ python
Python 3.8.8 (default, Apr 13 2021, 12:59:45) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> print(sys.version)
3.8.8 (default, Apr 13 2021, 12:59:45) 
[Clang 10.0.0 ]
>>> exit()

I don't know how to get more information. Does this help?

**Reply by v-chojas** [reply]
```
import platform
print(platform.platform())
```

**Reply by sm-ffm** [reply]
> macOS-10.16-x86_64-i386-64bit

**Reply by v-chojas** [reply, marked as answer]
There's your problem: you're running x86 Python on an ARM64 system, and attempting to use it with an ARM64 ODBC driver.
Install an ARM64 Python and try again.

**Reply by sm-ffm** [reply]
Thanks a lot! It worked!
